src/controllers/bulk/bulkcontroller.h

```diff
@@ -72,13 +72,12 @@ class BulkController : public Controller {
 
     // Local copies of things we need from desc
 
-    unsigned short m_vendorId;
-    unsigned short m_productId;
-    unsigned char m_inEndpointAddr;
-    unsigned char m_outEndpointAddr;
-#if defined(__WINDOWS__) || defined(__APPLE__)
-    unsigned int m_interfaceNumber;
-#endif
+    std::uint16_t m_vendorId;
+    std::uint16_t m_productId;
+    std::uint8_t m_inEndpointAddr;
+    std::uint8_t m_outEndpointAddr;
+    std::uint8_t m_interfaceNumber;
```

Comment:
This should become an int, to match the usbapi type and to allow us to retain the 2.5 behaviour for the devices where we don't know the interface number by setting it to -1.

Reply:
std::optional<std::uint8_t> would be so much more appropriate though.

Reply:
Yes, when starting on a plan surface. 

Libusb has decided to use -1 for invalid, so we should IMHO not introduce another concept. The API will just work with -1 and has internal checks, no need for any has_value() checks, in our code.

Reply:
> Libusb has decided to use -1 for invalid,

I was not able to find that definition. Can you link it to me?

> The API will just work with -1 and has internal checks, no need for any has_value() checks, in our code.

Well, if we want to preserve how it works currently though, we need to check for -1 anyways, so we might as well check for has_value. I'm not sure if passing -1 deliberately is a good choice because it'll only make it more complicated to then handle the error (because we need to then distinguish from an error that occurred because we passed in -1 or we actually did pass in a garbage value we didn't account for).

Reply:
It is just used literally in libusb like here: 
https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/descriptor.c#L221
and 
https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/os/windows_winusb.c#L2927

The only related macro is USB_MAXINTERFACES

Reply:
The change has been introduced here: 
https://github.com/libusb/libusb/commit/e9eec3a680cad3b2c9c5213fb7d60148cf6900da

All function taking int interface number and are checked for < 0 and >= USB_MAXINTERFACES so -1 is just one particular invalid interface number. 

There is: 
https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/os/windows_winusb.c#L2998-L3005

In our code is: 
```
constexpr static int kInvalidInterfaceNumber = -1;
```

That are my point voting for -1. If you still like to use an optional `std::optional<std::uint8_t>` and introduce extra checks and conversions to int, go ahead.

Reply:
No checks or conversions are needed, because the case of an invalid interface number is impossible here. As the hardcoded lookup table in Mixxx is the sole source of this information, and libusb the receiver.

I think you are confusing this with hidapi, which allows you to use transport protocols other than USB. And for these alternative transport protocols, there can't exist a valid USB interface number.

Reply:
From what I could tell is that @daschauer wants to introduce a sentinel value (-1 currently) that determines whether we should claim the interface at all. At least the current code only claimed an interface when it was non-zero. Whether that is correct is dubious so I just implemented it to claim it unconditionally.

Reply:
Anyways, I think we should try the state as currently implemented. None of us has a hercules controller anyways so its not like we can easily verify that this breaks something.

Reply:
It is impossible to transfer data to or from an USB device without specifiing the interface (there might be a default, but it must be specified somehow). This is mandatory by the standard.
But hardcoding the Endpoint Addresses is not in sense of the standard, as these should be read back from the descriptors in the device, taking into account the currently active device configuration and alternate mode of the interface .

Reply:
Claiming interfaces has been introduced recently in this PR https://github.com/mixxxdj/mixxx/pull/13008. 
The code must have been worked before at least on Linux without claiming the interfaces. I have read that the Kernel automatically detaches the driver and frees up the interface for the user space application without explicit claiming.

The libusb implementation is implemented in a way that a handle can claim one interface or more. Transferring data works with the endpoint address only, without specifying the interface. So I assume that the endpoint address alone is sufficient for transferring the data. Claiming the interface seems to be only an exclusive access thing. 

Searching the forum has revealed that the interface number is not always 0. 

> It is impossible to transfer data to or from an USB device without specifying the interface

I am afraid this is wrong. My conclusion is that claiming the interface 0 here for all devices is also wrong, since we don't have the knowledge which interface to claim. As a workaround we may restore the pre https://github.com/mixxxdj/mixxx/pull/13008 by not claiming the interface indicated by `-1` or `std::nullopt` if you wish.
So we may call the variable:  
`kClaimNoInterface = -1;`

Reply:
I agree, if we don't know the correct interface number, we shouldn't assume that the device use 0.